include/RAJA/pattern/kernel/ArgHelper.hpp

```diff
@@ -0,0 +1,293 @@
+ */
+
+
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2016-19, Lawrence Livermore National Security, LLC.
+//
+// Produced at the Lawrence Livermore National Laboratory
+//
+// LLNL-CODE-689114
+//
+// All rights reserved.
+//
+// This file is part of RAJA.
+//
+// For details about use and distribution, please read RAJA/LICENSE.
+//
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+
+#ifndef RAJA_pattern_kernel_arghelper_HPP
+#define RAJA_pattern_kernel_arghelper_HPP
+
+#include "RAJA/config.hpp"
+
+#include <iostream>
+#include <type_traits>
+
+#include "camp/camp.hpp"
+#include "camp/concepts.hpp"
+#include "camp/tuple.hpp"
+
+#include "RAJA/util/macros.hpp"
+#include "RAJA/util/types.hpp"
+
+#include "RAJA/pattern/kernel/OffSet.hpp"
+#include "RAJA/pattern/kernel/Param.hpp"
+#include "RAJA/pattern/kernel/Seg.hpp"
+
+namespace RAJA
+{
+
+namespace internal
+{
+
+using RAJA::statement::SegList;
+using RAJA::statement::Seg;
+
+using RAJA::statement::OffSet;
+using RAJA::statement::OffSetList;
+
+using RAJA::statement::ParamList;
+using RAJA::statement::Param;
+
+//Extracts arguments from segments, and parameters
+template<typename Head, typename...Tail>
+struct extractor
+{};
+
+template<camp::idx_t id, typename...Tail>
+struct extractor<RAJA::statement::OffSet<id>, Tail...>
+{
+  template<typename Data>
+  RAJA_HOST_DEVICE
+  static auto extract_arg(Data &&data) ->
+    decltype(camp::get<id>(data.offset_tuple))
+  {
+    return camp::get<id>(data.offset_tuple);
+  }
+};
+
+template<camp::idx_t id, typename...Tail>
+struct extractor<RAJA::statement::Seg<id>, Tail...>
+{
+  template<typename Data>
+  RAJA_HOST_DEVICE
+  static auto extract_arg(Data &&data) ->
+    decltype(camp::get<id>(data.segment_tuple).begin()[camp::get<id>(data.offset_tuple)])
+  {
+    return camp::get<id>(data.segment_tuple).begin()[camp::get<id>(data.offset_tuple)];
+  }
+};
+
+template<camp::idx_t id, typename...Tail>
+struct extractor<RAJA::statement::Param<id>, Tail...>
+{
+  template<typename Data>
+  RAJA_HOST_DEVICE
+  static auto extract_arg(Data &&data)->
+    decltype(camp::get<id>(data.param_tuple))
+  {
+    return camp::get<id>(data.param_tuple);
+  }
+};
+
+//Helper to concatante list [Move to camp?]
+template<typename ListA, typename ListB>
+struct catList
+{};
+
+template<typename...itemsA, typename...itemsB>
+struct catList<camp::list<itemsA...>, camp::list<itemsB...>> {
+
+  RAJA_HOST_DEVICE
+  static auto makeList(camp::list<itemsA...> const &, camp::list<itemsB...> const &) ->
+    camp::list<itemsA...,itemsB...>
+  {
+    return camp::list<itemsA...,itemsB...> {};
+  }
+
+};
+
+template<typename...itemsA>
+struct catList<camp::list<itemsA...>, camp::list<>> {
+
+  RAJA_HOST_DEVICE
+  static auto makeList(camp::list<itemsA...> const &, camp::list<> const &) ->
+    camp::list<itemsA...>
+  {
+    return camp::list<itemsA...> {};
+  }
+
+};
+
+template<typename Arg>
```

Comment:
Same comment for listMaker as for catList... convert to type based calculation

Reply:
Maybe we should chat about some of this, I have a feeling there are camp components that could slim this down quite a bit.  Hazard of the fact I haven't documented camp well, need to make some time for that.

Reply:
@trws , should I send an invite for this week? Or feel free to prose something via email.